### PR TITLE
Fix studio agent flaky test

### PR DESCRIPTION
**Pull request by oliversun9**
Tests `TestPlainPostHandlerTLS/invalid_upstream` and `TestPlainPostHandlerH2C/invalid_upstream` are [flaky](https://github.com/bufbuild/buf/actions/runs/9599996622/job/26475130640?pr=3102), because the listener only calls `Accept` once, and when it accepts a connection from something other than the agent, the listener calls it a day, leaving the agent hanging.

This PR makes the listener repeatedly accept new connections until it has accepted a connection from the agent.

Running `go test -count=1 -timeout=20s -race github.com/bufbuild/buf/private/buf/bufstudioagent` on main on my laptop would fail (with timeout) this test within the first 5 times, and trying it on this branch, it doesn't fail (tried 100 times and it hasn't failed).

TLDR: random ports can make connection with the listener and we need to account for that.

### CODE REVIEW

File: private/buf/bufstudioagent/bufstudioagent_test.go [hunk 1]
```diff
@@ -193,14 +194,31 @@ func testPlainPostHandlerErrors(t *testing.T, upstreamServer *httptest.Server) {
...
-			conn, err := listener.Accept()
-			require.NoError(t, err)
-			require.NoError(t, conn.Close())
+			for {
+				select {
+				case <-done:
+					listener.Close()
+					return
+				default:
+					err := listener.SetDeadline(time.Now().Add(time.Second))
```
Comment (oliversun9) [inline private/buf/bufstudioagent/bufstudioagent_test.go:212]: `time.Second` is an arbitrary number, but the nice thing about it is that when you run `go test -count=1 -timeout=20s -race github.com/bufbuild/buf/private/buf/bufstudioagent` you can estimate how many connections did it accept before accepting the one coming from the studio agent.

Comment (doriable) [reply private/buf/bufstudioagent/bufstudioagent_test.go:212]: I'm okay with leaving this deadline for now, but for testing, I think it's fine to retry pretty aggressively (e.g. `time.Milisecond`), since we want the test to run as quickly as possible.

**Review by doriable — APPROVED** [review]
Looks good to me, thanks for fixing this up!

File: private/buf/bufstudioagent/bufstudioagent_test.go [hunk 2]
```diff
@@ -193,14 +194,31 @@ func testPlainPostHandlerErrors(t *testing.T, upstreamServer *httptest.Server) {
...
+			done <- struct{}{}
+		})
 		go func() {
-			conn, err := listener.Accept()
-			require.NoError(t, err)
-			require.NoError(t, conn.Close())
+			for {
+				select {
+				case <-done:
+					listener.Close()
```
Comment (emcfarlane) [inline private/buf/bufstudioagent/bufstudioagent_test.go:209]: Why the channel? It's equivalent to call `listener.Close()` in the Cleanup and you already must handle the error on accept.

File: private/buf/bufstudioagent/bufstudioagent_test.go [hunk 3]
```diff
@@ -193,14 +194,31 @@ func testPlainPostHandlerErrors(t *testing.T, upstreamServer *httptest.Server) {
...
+					return
+				default:
+					err := listener.SetDeadline(time.Now().Add(time.Second))
+					require.NoError(t, err)
+					conn, err := listener.Accept()
+					if err != nil {
+						continue
+					}
+					require.NoError(t, conn.Close())
```
Comment (emcfarlane) [inline private/buf/bufstudioagent/bufstudioagent_test.go:218]: `require` is not safe from go routines: https://pkg.go.dev/testing#T.FailNow

Comment (oliversun9) [reply private/buf/bufstudioagent/bufstudioagent_test.go:218]: Good catch!

File: private/buf/bufstudioagent/bufstudioagent_test.go (hunk 2 above)
Comment (oliversun9) [reply private/buf/bufstudioagent/bufstudioagent_test.go:209]: If I call `listener.Close()` from the parent goroutine, the listener wouldn't know it's closed (plz correct me if I'm wrong) and then would just keep calling `Accept` even though it's already closed.

Comment (emcfarlane) [reply private/buf/bufstudioagent/bufstudioagent_test.go:209]: Ah the continue would need to do error handling. Maybe this is fine.

File: private/buf/bufstudioagent/bufstudioagent_test.go (hunk 3 above)
Comment (oliversun9) [reply private/buf/bufstudioagent/bufstudioagent_test.go]: Updated them to `continue`

File: private/buf/bufstudioagent/bufstudioagent_test.go (hunk 1 above)
Comment (oliversun9) [reply private/buf/bufstudioagent/bufstudioagent_test.go]: Updated to 10ms: [dde47d6](https://github.com/bufbuild/buf/pull/3113/commits/dde47d68dced33e354b66dcbb698f4cf2f0f22d0), I'm not sure how what would happen with 1ms when GitHub action runners' performance is degraded.

File: private/buf/bufstudioagent/bufstudioagent_test.go (hunk 3 above)
Comment (emcfarlane) [reply private/buf/bufstudioagent/bufstudioagent_test.go]: You can use `t.Error` safely as that will call fail not failnow.

Comment (oliversun9) [reply private/buf/bufstudioagent/bufstudioagent_test.go]: Fixed in [4852e27](https://github.com/bufbuild/buf/pull/3113/commits/4852e27e51e59999c54730a4eb23f841423474c4)

**Review by bufdev — APPROVED** [review]
